File: daily/Backtracking/131-Palindrome-partitioning.py
class Solution:
    def partition(self, s: str) -> List[List[str]]:
        res = []
        def is_palindrome(string) -> bool:
            left, right = 0, len(string) - 1
            while left < right:
                if string[left] != string[right]:
                    return False
                left += 1
                right -= 1
            return True
        def backtrack(start, path):
            if start == len(s):
                res.append(path[:])
                return
            for end in range(start + 1, len(s) + 1):
                sub_string = s[start:end]
                if is_palindrome(sub_string):
                    path.append(sub_string)
                    backtrack(end, path)
                    path.pop()
        backtrack(0, [])
        return res

File: daily/Backtracking/131-Palindrome-partitioning.go
package daily

func partition(s string) [][]string {
	var res [][]string
	var path []string

	isPal := func(sub string) bool {
		l, r := 0, len(sub)-1
		for l < r {
			if sub[l] != sub[r] {
				return false
			}
			l++
			r--
		}
		return true
	}

	var backtrack func(start int)
	backtrack = func(start int) {
		if start == len(s) {
			part := make([]string, len(path))
			copy(part, path)
			res = append(res, part)
			return
		}
		for end := start + 1; end <= len(s); end++ {
			sub := s[start:end]
			if isPal(sub) {
				path = append(path, sub)
				backtrack(end)
				path = path[:len(path)-1]
			}
		}
	}

	backtrack(0)
	return res
}
